Change TA tables to use declarative partitioning and other enhancements

chmnata opened the pull request:
## What this pull request accomplishes:

Restructure `here.ta` tables:
- Change from partitioning using inheritance to declarative partitioning 
- Separate column `tx` (timestamp) to `dt` (date) and `tod` (time)
- Change insert triggers to go through simple view

## Issue(s) this solves:

- Closes #494 
- Closes #495 
- Closes #496 
- Closes #499 

## What, in particular, needs to reviewed:
- `create_table.sql` :  new parent table structure + partitioned table creation + simple view + trigger on view
- `move_table.sql`: the loop to move old data to new data tables
- `function_create_yearly_table.sql`: function to create yearly partition tables
- `trigger_here_insert.sql`: trigger to insert data to parent table through simple view

Question: 
- Having some problem understanding whether there is a need on putting index on `dt` as it is the partition column. Under 5.10.4.  Partitioning and Constraint Exclusion. I'm wondering if we need to add it when a lot of the times we are filtering for months and not specific dates 🤔 🤔 🤔 Maybe Dr. Doc can comment on this :cat_doctor:
>Note that constraint exclusion is driven only by CHECK constraints, not by the presence of indexes. Therefore it isn't necessary to define indexes on the key columns. Whether an index needs to be created for a given partition depends on whether you expect that queries that scan the partition will generally scan a large part of the partition or just a small part. An index will be helpful in the latter case but not the former.
https://www.postgresql.org/docs/10/ddl-partitioning.html


Notes:
- There is no `ON CONFLICT DO NOTHING` for the insert trigger because we can't put that for parent table in PG version 10 or lower (can do in 11). There are unique constraints on tod, dt, and link_dir for each partitioned table to prevent inserting duplicated data.
- Cannot put indexes on parent table because our PG version 10 or lower, can be changed in 11. So there are still indexes created for each partitioned table.

radumas commented on here/traffic/sql/create_tables.sql:
```diff
@@ -1,73 +1,72 @@
+-- simple view structure
+CREATE VIEW here.ta_view AS
+
+SELECT *
+FROM here.ta;
```
This needs to exclude your transformed columns (`dt` and `tod`)

radumas replied:
This needs to exclude the transformed columns `dt` and `tod`

radumas commented on here/traffic/sql/create_tables.sql:
```diff
@@ -1,73 +1,72 @@
 
+
+--Loops through the years and then months for which we currently have data in order to create a partitioned table for each month
 DO $do$
```
Instead of doing this looping you could loop over that yearly create tables function.

Or use the eoy DAG and run it for 2012 to 2022

radumas commented on here/traffic/sql/create_tables.sql:
```diff
@@ -1,73 +1,72 @@
-pct_35 int NOT NULL,
-pct_40 int NOT NULL,
-pct_45 int NOT NULL,
-pct_50 int NOT NULL,
```
I think the median speed might be worth keeping.

radumas commented on here/traffic/sql/create_tables.sql:
```diff
@@ -1,73 +1,72 @@
-pct_80 int NOT NULL,
-pct_85 int NOT NULL,
-pct_90 int NOT NULL,
-pct_95 int NOT NULL
```
from your speeding analysis of HERE, are any of the upper percentiles worth keeping?

chmnata replied:
ahhhhhhhhhhhhhhhhhhh yes, seems like we used pct_50, pct_85, and the max. I will include those.

radumas commented on here/traffic/sql/move_tables.sql:
```diff
@@ -0,0 +1,35 @@
+-- Move data from here.ta_* to the newly structured here.ta_*
+-- while performing separation of tx column to dt (Date) and tod (time)
+-- loop through each table, move data to new table, then truncate the old one
+--Loops through the years and then months for which we currently have data in order to create a partitioned table for each month
+
```
This needs to be done in Python, because otherwise committing won't happen until the loop is finished and it's going to be a big deal

radumas commented on here/traffic/sql/trigger_here_insert.sql:
```diff
@@ -0,0 +1,22 @@
+							NEW.min_spd,
+							NEW.max_spd,
+							NEW.confidence,
+                            NEW.sample_size); -- postgresql <=10 does not support ON CONFLICT DO NOTHING
```
Can you link to this?

chmnata replied:
Under 5.10.2.3. Limitations
> Using the ON CONFLICT clause with partitioned tables will cause an error, because unique or exclusion constraints can only be created on individual partitions. There is no support for enforcing uniqueness (or an exclusion constraint) across an entire partitioning hierarchy.

I'm guessing we will have to create additional triggers for each partitioned table if `ON CONFLICT` is needed.

radumas replied:
ah shoot well I guess at least insert will fail if trying to insert duplicate data, which is fine. It's not like other pipelines where we might get partial data

radumas reviewed (changes requested):
Great work documenting this PR.

Can you write out the transition plan you have for moving over the data?

>Having some problem understanding whether there is a need on putting index on dt as it is the partition column. Under 5.10.4. Partitioning and Constraint Exclusion. I'm wondering if we need to add it when a lot of the times we are filtering for months and not specific dates

But you did create an index on this column for the child tables. I think that's what the documentation is also discussing, whether it's worthwhile to put an index on the children tables. `constraint_exclusion` will quickly determine which tables to query based on the requested dates, but we still need an index to quickly retrieve rows within that table.

Some other comments on the individual elements

chmnata commented:
## Transition plan

1) Create `here.ta_new` with declarative partitioning. Because it is not possible to alter `here.ta` (technically a regular table) to a partitioned table.
> It is not possible to turn a regular table into a partitioned table or vice versa. However, it is possible to add an existing regular or partitioned table as a partition of a partitioned table, or remove a partition from a partitioned table turning it into a standalone table; this can simplify and speed up many maintenance processes. See [ALTER TABLE](https://www.postgresql.org/docs/current/sql-altertable.html) to learn more about the ATTACH PARTITION and DETACH PARTITION sub-commands.
2) Create a new table with all transformation using old child table for every month.
```
CREATE TABLE here.ta_201201_new AS
SELECT columns
FROM here.ta_201201
```
3) Attach newly created table as a partition child table to the newly created declarative parent table. Adding a check because it allow the system to skip the scan when validating the partition constraint. see below text. (Step 2 and 3 will be in a loop in a .py`
```
ALTER TABLE here.ta_201201_new ADD CONSTRAINT y2012m01
   CHECK ( dt>= DATE '2012-01-01' AND dt< DATE '2012-02-01' );
   
ALTER TABLE here.ta_201201_new ATTACH PARTITION here.ta_new
    FOR VALUES FROM ('2012-01-01') TO ('2012-02-01' );
  
DROP CONSTRAINT y2012m01;
```
>Before running the ATTACH PARTITION command, it is recommended to create a CHECK constraint on the table to be attached matching the desired partition constraint. That way, the system will be able to skip the scan to validate the implicit partition constraint. Without the CHECK constraint, the table will be scanned to validate the partition constraint while holding an ACCESS EXCLUSIVE lock on the parent table. It may be desired to drop the redundant CHECK constraint after ATTACH PARTITION is finished.

4) Change all dependencies of `here.ta` to `here.ta_new`. @radumas Most of them are in `data_requests` schema, I think we can just drop those without recreating them since they should all be committed anyways. Some are the reporting mat views for rapidto. Some are in our personal schemas. Will make a list and we can decide on what to do with each of them.

5) Drop `here.ta`, and make sure there are no other hidden dependencies

6) Rename `here.ta_new` to `here.ta`

7) Pray everything works 🙏🏼

radumas commented on here/traffic/move_monthly_table.py:
```diff
@@ -0,0 +1,28 @@
+logging.basicConfig(level=logging.INFO, format=FORMAT)
+LOGGER = logging.getLogger(__name__)
+
+yyyy_list = ['2012', '2013', '2014', '2015', '2016', '2017', '2018']
```
shouldn't this go up to 2017-09?

radumas commented on here/traffic/move_monthly_table.py:
```diff
@@ -0,0 +1,28 @@
+        LOGGER.info('Processing month %s of year %s', m, y)
+        with con:
+            with con.cursor() as cur:
+                cur.execute("SELECT here.move_data(%s, %s);", (y,m))
```
`here_staging.move_data`

radumas commented on here/traffic/sql/transition.sql:
```diff
@@ -0,0 +1,105 @@
+-- Create view in staging schema
+CREATE VIEW here_staging.ta_view AS
+ 
+select link_dir, tx, length, mean, stddev, min_spd, max_spd, pct_50, pct_85, confidence, sample_size 
```
does this need an epoch place-holder?

chmnata commented:
https://github.com/CityofToronto/bdit_data-sources/blob/0deda9c35cca500bf0b85b72cce7850a6a22a8a9/here/traffic/here_api.py#L65
I feel like we don't specify the mapversion 🤔 and the default is 18_3

chmnata commented:
_Update on PR as of 2022-08-24_
## What this pull request accomplishes:

Restructure `here.ta` tables:
- Change from partitioning using inheritance to declarative partitioning 
- Separate column `tx` (timestamp) to `dt` (date) and `tod` (time)
- Added `sample_size` column
- Change insert triggers to go through simple view

Updates on here_api:
- Direct COPY to `here.ta_view` instead of `here.ta`
- Modified request query parameters to include sample_size and exclude pct_x other than pct_50 and pct_85
- Attempted to catch Key Error (#500)

## What, in particular, needs to reviewed:
- `create_table.sql` :  new parent table structure + partitioned table creation + simple view + trigger on view
- `trigger_here_insert.sql`: trigger to insert data to parent table through simple view
- `here_api.py` : Python script to pull here ta data from here api (mainly the query part and error catching portion)

radumas commented on here/traffic/here_api.py:
```diff
@@ -47,6 +47,20 @@ def get_access_token(key_id, key_secret, token_url):
+        finally:
+            raise HereAPIException(err_msg)
+
     access_token = r.json()['accessToken']
```
this line should be in the try block

chmnata replied:
should it be catching different errors than line 50 to 62? Just Jsondecodeerror? and maybe another keyerror?

radumas commented on here/traffic/here_api.py:
```diff
@@ -47,6 +47,20 @@ def get_access_token(key_id, key_secret, token_url):
+    try:
+        r.raise_for_status()
+    except KeyError as k_err:
+        LOGGER.error(k_err)
```
This should be 
```python
LOGGER.error('Key Error in getting access token, response was')
raise HereAPIException(r.text)
```

radumas commented on here/traffic/here_api.py:
```diff
@@ -108,6 +122,7 @@ def get_download_url(request_id, status_base_url, access_token, user_id):
         LOGGER.info('Polling status of query request: %s', request_id)
         query_status = requests.get(status_url, headers = status_header)
         try:
+            query_status.raise_for_status()
```
Should probably include the same `except requests.exceptions.HTTPError as err:` from above

radumas commented on here/traffic/here_api.py:
```diff
@@ -165,7 +182,7 @@ def send_data_to_database(ctx=None, datafile = None, dbsetting=None):
         #Second uses check_call and 'ON_ERROR_STOP=1' to make sure errors are captured and that the third 
         #process doesn't run befor psql is finished.
-        copy = r'''"\COPY here.ta FROM STDIN WITH (FORMAT csv, HEADER 
+        copy = r'''"\COPY here_staging.ta_view FROM STDIN WITH (FORMAT csv, HEADER 
```
should this be `here.ta_view` after transition is complete?

radumas commented on here/traffic/sql/create_tables.sql:
```diff
@@ -1,73 +1,53 @@
+        tx, 
+        NULL:integer as epoch_min, 
+        ength, 
+        sample_size, 
```
flagging that sample size is not in the same location as table. Confirm this is correct location

chmnata replied:
yup! confirming that this is the correct column order for the input csv.gz to copy to. Then the trigger reorganize the columns to fit the here.ta table column order.

radumas reviewed (changes requested):
Great work. a few edits in the error handling in `here_api.py`.

SQL looks fine just confirm that sample_size is in the correct location in the view vs. the table.

radumas reviewed (changes requested):
sql is good. Some problems in the python file.

radumas commented on here/traffic/here_api.py:
```diff
@@ -51,7 +51,13 @@ def get_access_token(key_id, key_secret, token_url):
+        except JSONDecodeError:
+            err_msg = r.text
+        finally:
+            LOGGER.error(err_msg)
```
stil need `raise HereAPIException(r.text)` otherwise the script won't actually fail properly here.

radumas commented on here/traffic/here_api.py:
```diff
@@ -130,7 +136,7 @@ def get_download_url(request_id, status_base_url, access_token, user_id):
         LOGGER.info('Polling status of query request: %s', request_id)
         query_status = requests.get(status_url, headers = status_header)
         try:
-            query_status.raise_for_status()
```
still wanted this line in

radumas commented on here/traffic/here_api.py:
```diff
@@ -137,6 +138,7 @@ def get_download_url(request_id, status_base_url, access_token, user_id):
         query_status = requests.get(status_url, headers = status_header)
         try:
             status = str(query_status.json()['status'])
+            query_status.raise_for_status()
```
these lines are in the wrong order.

radumas reviewed (changes requested):
Simplifying the error catching code for the two functions parsing responses from the API. Hopefully this... works.

for the getting access token:

```python
    try:
        r.raise_for_status()
        access_token = r.json()['accessToken']
    except (requests.exceptions.HTTPError, KeyError, JSONDecodeError, ValueError)  as err:
        error = 'Error in requesting access token \n'
        error += 'Response was:\n'
        try:
            resp = r.json()['message']
        except JSONDecodeError:
            resp = r.text
        finally:
            error += resp
            error += '\n'
            raise HereAPIException(error)
```

For the get download URL

```python
        try:
            query_status.raise_for_status()
            status = str(query_status.json()['status'])
        except (requests.exceptions.HTTPError, KeyError) as err:
            error = 'Error in polling status of query request \n'
            error += 'err\n'
            error += 'Response was:\n'
            try:
                resp = str(query_status.json()['message'])
            except JSONDecodeError:
                resp = query_status.text
            finally:
                error += resp
                error += '\n'
                raise HereAPIException(error)
        except JSONDecodeError as json_err:
            LOGGER.warning("JSON error in query status response.")
            LOGGER.warning(query_status.text)
            continue
```

radumas commented:
@chmnata CodeFactor raised an smol ❌

radumas reviewed (approved):
My name is @radumas and I approve this Pull Request :here: :octocat: :meow_praise:
